### ISSUE TITLE
Mounts without a name strips the path arg from calls to child routes in url_path_for

### ISSUE THREAD

**Issue by jsatt** [Kludex/starlette#687]
I'm using an unnamed Mount which contains a mount to a StaticFiles instance and when using "url_for" it is unable to find the matching file. After a bit of digging it seems to be directly related to url_for_path in Mount clearing the path arg whenever the mount does not have a name set. 

https://github.com/encode/starlette/blob/master/starlette/routing.py#L337

Calling url_path_for in the child router works as expected. I'm unclear as to why you'd want to strip this arg just because the mount is not named.

```
app.mount('/stuff', app=Router(routes=[
    Mount('/static', app=StaticFiles(directory='static'), name='static'),
]))
```

```
In [6]: app.url_path_for('static', path='css/bootstap.min.css')
---------------------------------------------------------------------------
NoMatchFound
...
In [7]: app.routes[0].routes[0].url_path_for('static', path='css/bootstap.min.css')
Out[7]: '/static/css/bootstap.min.css'
```

**Comment by mancubus77** [comment]
+1 For the issue
Can not getting work `url_for` for Route()
```
app = Router([
    Route('/', endpoint=Homepage, methods=['GET'])
    ])
```
```
  File "templates/base.html", line 13, in top-level template code
    <link href="{{ url_for('static', path='/css/bootstrap.min.css') }}" rel="stylesheet">
...
  File "/home/stan/devel/starlette-example/venv/lib/python3.6/site-packages/starlette/routing.py", line 576, in url_path_for
    raise NoMatchFound()
starlette.routing.NoMatchFound
```
```
>>> print(starlette.__version__)
0.12.10
```

**Comment by VeNoMouS** [comment]
I have the exact same issue with the same example.. https://github.com/encode/starlette-example

**Comment by gvbgduh** [comment]
@tomchristie would it be reasonable to consider making it a bit stricter, like
```
(
    (
        '/stuff', (
            '/static', ..., name='static',
            '/thing', ..., name='thing'
        ),
        name = None
    ),
    (
        '/staff', (
            '/{id}', (
                '/details', ..., 'details',
            ), name='person',
        ),
        name = 'staff'
    ),
    (
        '/products', (
            '/{id}', (
                '/details', ..., 'details',
            ), name='item',
        ),
        name = 'products'
    )
)
```
so, it'll be a tree (shown as a dict for simplicity)
```
{
    'staff': {
        'person': {
            'details': None
        },
    }
    'products': {
        'item': {
            'details': None
        }
    }
}
```
and look up for
- `staff:person:details` returns `/staff/{id}/details`
- `products:brand:details` returns exception like _No registered name "brand" for "product"_

And for the initial issue statements: `static` -> _No registered name "static"_

**Comment by rmorison** [comment]
+1, spent 2h debugging before I thought to check repo:issues. Documentation is also lacking on how `name`s compose in nested routing for `url_for`. This thread has more detail than the docs.

**Comment by lovelydinosaur** [comment]
Thanks for raising this. There's a few seperate issues here:

* `url_for` was broken for the "Mount within Mount" case. Resolved in #701.
* The docs are poor.
* The raised exceptions are not helpful.

I'm going to spend some more comprehensive time on this next week.

I'm also planning to start switching our docs out to use the routing table style, instead of the decorator style eg...

```python
routes = [
    Route('/', homepage),
    Mount('/static', app=StaticFiles(directory='static'), name='static'),
]

app = Starlette(routes=routes)
```

I think it's a whole bunch cleaner & works better as projects grow.

That won't mean getting rid of the decorator and `add_route` APIs, but it will mean changing the docs so that we're pushing for that particular style, and sidelining the alternatives.
